File: clients/maestro_tui/lib/the_maestro/tools/seek_sequence.ex
defmodule TheMaestro.Tools.SeekSequence do
  @moduledoc false

  @spec seek_sequence([String.t()], [String.t()], non_neg_integer(), boolean()) :: non_neg_integer() | nil
  def seek_sequence(lines, pattern, start, eof?) do
    cond do
      pattern == [] -> start
      length(pattern) > length(lines) -> nil
      true ->
        search_start = if eof? and length(lines) >= length(pattern), do: length(lines) - length(pattern), else: start
        exact(lines, pattern, search_start) || rstrip(lines, pattern, search_start) || trim(lines, pattern, search_start) || normalized(lines, pattern, search_start)
    end
  end

  defp exact(lines, pattern, start) do
    last = length(lines) - length(pattern)
    Enum.find(0..max(last - start, 0), fn off -> Enum.slice(lines, start + off, length(pattern)) == pattern end)
    |> case do nil -> nil; off -> start + off end
  end

  defp rstrip(lines, pattern, start) do
    last = length(lines) - length(pattern)
    Enum.find(0..max(last - start, 0), fn off -> Enum.zip(Enum.slice(lines, start + off, length(pattern)), pattern) |> Enum.all?(fn {a, b} -> String.trim_trailing(a) == String.trim_trailing(b) end) end)
    |> case do nil -> nil; off -> start + off end
  end

  defp trim(lines, pattern, start) do
    last = length(lines) - length(pattern)
    Enum.find(0..max(last - start, 0), fn off -> Enum.zip(Enum.slice(lines, start + off, length(pattern)), pattern) |> Enum.all?(fn {a, b} -> String.trim(a) == String.trim(b) end) end)
    |> case do nil -> nil; off -> start + off end
  end

  defp normalized(lines, pattern, start) do
    norm = fn s ->
      s
      |> String.trim()
      |> String.graphemes()
      |> Enum.map(fn
        <<?–::utf8>> -> "-"
        <<?—::utf8>> -> "-"
        <<?−::utf8>> -> "-"
        <<?‑::utf8>> -> "-"
        <<?‘::utf8>> -> "'"
        <<?’::utf8>> -> "'"
        <<?“::utf8>> -> "\""
        <<?”::utf8>> -> "\""
        <<0xA0>> -> " "
        other -> other
      end)
      |> IO.iodata_to_binary()
    end

    last = length(lines) - length(pattern)
    Enum.find(0..max(last - start, 0), fn off -> Enum.zip(Enum.slice(lines, start + off, length(pattern)), pattern) |> Enum.all?(fn {a, b} -> norm.(a) == norm.(b) end) end)
    |> case do nil -> nil; off -> start + off end
  end
end

File: clients/maestro_tui_python/src/maestro_tui/tools/seek_sequence.py
from __future__ import annotations

from typing import List, Optional


def _norm_char(ch: str) -> str:
    if ch in {"–", "—", "−", "‑"}:
        return "-"
    if ch in {"‘", "’"}:
        return "'"
    if ch in {"“", "”"}:
        return '"'
    if ord(ch) == 0xA0:
        return " "
    return ch


def _normalize_line(s: str) -> str:
    s = s.strip()
    return "".join(_norm_char(c) for c in s)


def _find_exact(lines: List[str], pattern: List[str], start: int) -> Optional[int]:
    last = len(lines) - len(pattern)
    for off in range(max(last - start, 0) + 1):
        if lines[start + off : start + off + len(pattern)] == pattern:
            return start + off
    return None


def _find_rstrip(lines: List[str], pattern: List[str], start: int) -> Optional[int]:
    last = len(lines) - len(pattern)
    for off in range(max(last - start, 0) + 1):
        ok = True
        for a, b in zip(lines[start + off : start + off + len(pattern)], pattern):
            if a.rstrip() != b.rstrip():
                ok = False
                break
        if ok:
            return start + off
    return None


def _find_trim(lines: List[str], pattern: List[str], start: int) -> Optional[int]:
    last = len(lines) - len(pattern)
    for off in range(max(last - start, 0) + 1):
        ok = True
        for a, b in zip(lines[start + off : start + off + len(pattern)], pattern):
            if a.strip() != b.strip():
                ok = False
                break
        if ok:
            return start + off
    return None


def _find_normalized(lines: List[str], pattern: List[str], start: int) -> Optional[int]:
    last = len(lines) - len(pattern)
    pat = [_normalize_line(p) for p in pattern]
    for off in range(max(last - start, 0) + 1):
        seg = [_normalize_line(a) for a in lines[start + off : start + off + len(pattern)]]
        if seg == pat:
            return start + off
    return None


def seek_sequence(lines: List[str], pattern: List[str], start: int, eof: bool) -> Optional[int]:
    if pattern == []:
        return start
    if len(pattern) > len(lines):
        return None
    search_start = len(lines) - len(pattern) if eof and len(lines) >= len(pattern) else start
    return (
        _find_exact(lines, pattern, search_start)
        or _find_rstrip(lines, pattern, search_start)
        or _find_trim(lines, pattern, search_start)
        or _find_normalized(lines, pattern, search_start)
    )

